### PR TITLE
Multiple steppers per axis on CoreXY

### PR DESCRIPTION
**Pull request by TypQxQ**
Changes and sample configuration to make it possible to use multiple steppers on +/- axis on a CoreXY.
Using a stepper in each corner shortens the slack of the belt in a CoreXY. This is especially useful in larger setups.

This is tested in a test setup because my current quad stepper CoreXY is running RRF and I'm waiting for a new board to test Klipper on it but works well in a test setup.

I know this was earlier asked for by the HevORT group.

Signed-off-by: Andrei Ignat <andrei@ignat.se>

### CODE REVIEW

File: config/sample-corexy-multi-mcu-multi-steppers.cfg [hunk 1]
```diff
@@ -0,0 +1,150 @@
+[include mainsail.cfg]
```
Comment (meteyou) [inline config/sample-corexy-multi-mcu-multi-steppers.cfg:1]: please remove the gui specific include. i think it should not be in a generic example config.

Comment (TypQxQ) [reply config/sample-corexy-multi-mcu-multi-steppers.cfg]: Thank you! Forgot to remove that.

**Comment by KevinOConnor** [comment]
Thanks.  I have the same comments as https://github.com/Klipper3d/klipper/pull/4884#issuecomment-1008502267 .  Additionally, I'd really like to get confirmation this works in practice, as I'm unsure of the interaction between the "multi printer rail" and the endstop assignment that is peculiar to corexy.

-Kevin

**Comment by TypQxQ** [comment]
I plan to test/migrate Klipper on the CoreXY I have set up with 2 steppers per belt. It is currently running RRF on Duet3 so I have to rewire it and I'm waiting for some connectors.

I have now deleted the example file. I thought that it might be useful as I have seen comments about it being hard to understand how to use multiple steppers with the current Cartesian config.

**Comment by TypQxQ** [comment]
> Thanks. I have the same comments as [#4884 (comment)](https://github.com/Klipper3d/klipper/pull/4884#issuecomment-1008502267) . Additionally, I'd really like to get confirmation this works in practice, as I'm unsure of the interaction between the "multi printer rail" and the endstop assignment that is peculiar to corexy.
> 
> -Kevin

Tested now and you were right. I had to change the endstop assignments. All steppers stop as expected on endstop now.

**Comment by Guillaume452** [comment]
this fork for klipper works very well, definately should be added to klipper main! For vzbot we have a new quad motor setup now, initially we thought just changing the first few lines in corexy.py to lookupmultirail would work but that left us (and also people in the hevort project) with a printer that would not home correctly (X1 and Y1 didnt stop when triggering endstops) even though the printer would print perfectly fine, really good actually.

 after someone discovered this branch, i tried this and works like a charm. I have tested this for a few days and have not seen any weird behaviour on the printer a

https://user-images.githubusercontent.com/93674339/155604949-0b597370-3e6e-4998-9b9f-14c45fa30ae2.mp4

nymore. 

In this youtube video i gave a little update on the upcoming release of VZ235 AWD and here it is seen printing and i explain the homing problem (wasn't solved back then). https://www.youtube.com/watch?v=1WG1JSY_2Ms&t=5s

i will attach a video of the printer homing correctly as wel.

We from the VzBot dev team and the people in the hevort community hope this will find its way into the klipper main branch!

**Comment by TypQxQ** [comment]
I also have a video of it working and homing: https://youtu.be/tFZND9To8cw

**Comment by JRHeaton** [comment]
This looks very promising and I am adding my +1 to support this. Great addition for larger CoreXY setups. I come from @TypQxQ's video.

**Comment by JRHeaton** [comment]
Perhaps what I would like to see before merging is documentation added, if needed, around the homing/endstop behavior with this setup. Just my 2 cents. Thanks.

**Comment by TypQxQ** [comment]
> Perhaps what I would like to see before merging is documentation added, if needed, around the homing/endstop behavior with this setup. Just my 2 cents. Thanks.

I thought so too and initially added a example config file but. See Kevins comments for reason it got removed.
Config is very easy. You can see my config files here:
https://github.com/TypQxQ/DuetBackup/tree/main/qTC-Klipper/klipper_config

**Comment by JRHeaton** [comment]
> > Perhaps what I would like to see before merging is documentation added, if needed, around the homing/endstop behavior with this setup. Just my 2 cents. Thanks.
> 
> I thought so too and initially added a example config file but. See Kevins comments for reason it got removed. Config is very easy. You can see my config files here: https://github.com/TypQxQ/DuetBackup/tree/main/qTC-Klipper/klipper_config

I don't know that everyone with this hardware setup is going to consider a 197 line "homing" file with macro replacements for core gcodes like `G28` "very easy." That doesn't necessarily feel natively supported. What am I missing? I admittedly haven't dived too deep into your whole config as I don't know what exactly applies to this besides the homing file.

**Comment by KevinOConnor** [comment]
Thanks.  At a high-level the change looks fine to me.  However, at the "merge level", this PR contains unrelated changes to stepper.py .  As near as I can tell, no changes are required to that file to make the corresponding update to corexy.py, and thus the two changes should not be mixed together.

-Kevin

**Comment by Guillaume452** [comment]
I would like to add that in the full release for quad motor vzbot, which will happen soon, i have added a full manual how to sync the belts, install this branch and how to configure in the cfg. Maybe something i can share for general documentation?

Aren't the changes in stepper.py for the homing? I had weird homing behaviout before i found this branch. Wouldnt it be possible to have the awd specific stuff in its own kinematic folder that will also alter the behaviour or affect the stepper.py file?

**Comment by TypQxQ** [comment]
@KevinOConnor: I have now removed the changes to stepper.py. They are only needed for Delta kinematics changes in #5217 .

**Comment by TypQxQ** [comment]
@Guillaume452 You can rename the corexy.py file in my fork to for example corexy_awd,py and in the printer.cfg, change to "kinematics: corexy_awd".

**Comment by Guillaume452** [comment]
> @Guillaume452 You can rename the corexy.py file in my fork to for example corexy_awd,py and in the printer.cfg, change to "kinematics: corexy_awd".

perfect, I will do that. did you test the changes to stepper.py allready? I'll test them this weekend myself too. will these changes make it so this can be merged to main klipper?:)

**Comment by KevinOConnor** [comment]
Thanks.

-Kevin

**Pull request merged**